Source/KerbalConstructionTime/Utilities/Utilities.cs

```diff
@@ -2122,6 +2126,131 @@ public static void CleanupDebris(string launchSiteName)
+
+        public static Tuple<float, List<string>> GetVesselUnlockInfo(ShipConstruct ship)
+        {
+            if (HighLogic.CurrentGame.Mode != Game.Modes.CAREER) return new Tuple<float, List<string>>(0, new List<string>());
+
+            // filter the ship parts list to those parts that are not already purchased
+            IEnumerable<KeyValuePair<AvailablePart, PartPurchasability>> purchasableParts = GetPartsWithPurchasability(ship.Parts).Where(kvp => kvp.Value.Status == PurchasabilityStatus.Purchasable || kvp.Value.Status == PurchasabilityStatus.Unavailable);
+            List<string> ecmPartsList = new List<string>();
+            float runningCost = 0;
+
+            // compare the part specified entry cost to the ECM database
+            foreach (AvailablePart part in purchasableParts.Select(kvp => kvp.Key))
+            {
+                int rawCost = part.entryCost;
+                double ecmEstCost = RealFuels.EntryCostManager.Instance.ConfigEntryCost(part.name);
+                if (rawCost == ecmEstCost)
+                {
+                    // this part is managed by the ECM, save its name later for a batch request
+                    string sanitizedName = RealFuels.Utilities.SanitizeName(part.name);
+                    if (!ecmPartsList.Contains(sanitizedName)) 
+                        ecmPartsList.Add(sanitizedName);
+                }
+                else
+                {
+                    // this part is not in the ECM, take the raw cost
+                    runningCost += rawCost;
+                }
+            }
+
+            // filter down further to those parts that can't be unlocked with our current tech and get the tech names needed
+            List<AvailablePart> lockedParts = purchasableParts.Where(kvp => kvp.Value.Status == PurchasabilityStatus.Unavailable).Select(kvp => kvp.Key).ToList();
+            HashSet<string> pendingTech = new HashSet<string>(lockedParts.Select(ap => ap.TechRequired));
+
+            // now back through the list looking for upgrades to add to our batch list
+            foreach (Part p in ship.Parts)
+            {
+                var mecs = p.FindModulesImplementing<RealFuels.ModuleEngineConfigsBase>();
+                foreach (RealFuels.ModuleEngineConfigsBase mec in mecs)
+                {
+                    string sanitizedName = RealFuels.Utilities.SanitizeName(mec.configuration);
+                    if (ecmPartsList.Contains(sanitizedName))
+                        continue;
+                    
+                    ecmPartsList.Add(sanitizedName);
+                    ConfigNode n = mec.configs.Find(c => c.GetValue("name") == mec.configuration);
+                    if (n != null)
+                    {
+                        string techRequired = n.GetValue("techRequired");
+                        if (techRequired == null) continue;
+
+                        ProtoTechNode techState = ResearchAndDevelopment.Instance.GetTechState(techRequired);
```

Comment:
You might want to consider just adding the techRequired for this (and for upgrades) to a hashset, and only at the end checking techstate for them all, to avoid duplication? Otherwise. Same with having ecmPartsList as a hashset. Otherwise you'll be doing a bunch of contains checks for every symmetry counterpart engine, etc.

Reply:
The contains check at the top of this block was to avoid doing the rest of the logic for duplicate parts. I think the contains there is cheaper than finding the config and then pulling it's tech repeatedly. I originally had it this way as I needed to know if it was a duplicate before checking costs, but even with the cost code no longer in this block, I think it's the faster approach.